File: benchmark/ff-gram-schmidt.pas
program FiniteFieldGramSchmidt;

{
    Finite field Gram-Schmidt: given n vectors of dimension d in F_p,
    finds an orthogonal basis of the subspace spanned by these n vectors.
    Here, p=1009.

    INPUT:
        line 1: space-separated d <= 1000, n <= 1000
        line 2..n+1: d space-separated residues modulo p

    OUTPUT:
        lines 1..n: d space-separated residues 0,..,p-1
        pairwise orthogonal, spanning the original subspace
        (some might be all-zeros if original vectors are not spanning)
}
const p = 1009; (* characteristic *)
var
    n, (* num of vectors *)
    d; (* dimension *)
    Mat: Array[0..999, 0..999] of integer;

(* compute a^e mod p *)
function Pow(a, e);
var b;
begin
    if e = 0 then
        Pow := 1
    else if e mod 2 = 1 then
        Pow := (a * Pow(a, e - 1)) mod p
    else
    begin
        b := Pow(a, e div 2);
        Pow := (b * b) mod p
    end
end;

(* compute dot product of ith and jth row, mod p *)
function Dot(i, j);
var k;
begin
    dot := 0;
    for k := 0 to d - 1 do
        dot := (dot + Mat[i, k] * Mat[j, k]) mod p;
end;

(* adds ith row to jth row, mod p *)
procedure Add(i, j);
var k;
begin
    for k := 0 to d - 1 do
        Mat[j, k] := (Mat[j, k] + Mat[i, k]) mod p;
end;

(* swap ith and jth row *)
procedure Swap(i, j);
var k, t;
begin
    for k := 0 to d - 1 do
    begin
        t := Mat[i, k];
        Mat[i, k] := Mat[j, k];
        Mat[j, k] := t
    end
end;

(*
    subtracts from jth vector the projection
    of the jth vector on to the ith vector,
    mod p
*)
procedure Orthogonalize(i, j);
var proj, k;
begin
    proj := (Dot(i, j) * Pow(Dot(i, i), p - 2)) mod p;
    for k := 0 to d - 1 do
        Mat[j, k] := (Mat[j, k] - ((proj * Mat[i, k]) mod p) + p) mod p;
end;

var k, i, j;
begin
    ReadLn(d, n);

    for k := 0 to n - 1 do
        for i := 0 to d - 1 do
            Read(Mat[k, i]);

    for k := 0 to n - 1 do
    begin
        i := k;
        while i < n do
        begin
            if Dot(i, i) <> 0 then
                break;
            i := i + 1
        end;

        if i < n then
            Swap(k, i)
        else
        begin
            i := k;
            while i < n do
            begin
                j := i + 1;
                while j < n do
                begin
                    if Dot(i, j) <> 0 then
                        break;
                    j := j + 1
                end;

                if j < n then
                    break;
                i := i + 1
            end;

            if i < n then
            begin
                Add(j, i);
                Swap(k, i)
            end
        end;

        if Dot(k, k) <> 0 then
        begin
            i := k + 1;
            while i < n do
            begin
                Orthogonalize(k, i);
                i := i + 1
            end
        end
    end;

    for k := 0 to n - 1 do
    begin
        for i := 0 to d - 1 do
            Write(Mat[k, i], ' ');
        WriteLn()
    end
end.

File: benchmark/ff-gram-schmidt.c
#include <stdio.h>

/*
    Finite field Gram-Schmidt: given n vectors of dimension d in F_p,
    finds an orthogonal basis of the subspace spanned by these n vectors.
    Here, p=1009.

    INPUT:
        line 1: space-separated d <= 1000, n <= 1000
        line 2..n+1: d space-separated residues modulo p

    OUTPUT:
        lines 1..n: d space-separated residues 0,..,p-1
        pairwise orthogonal, spanning the original subspace
        (some might be all-zeros if original vectors are not spanning)
*/

const int p = 1009;
int n, d;
int Mat[1000][1000];

int Pow(int a, int e) {
    int b;
    if (e == 0)
        return 1;
    else if (e % 2 == 1)
        return (a * Pow(a, e - 1)) % p;
    else {
        b = Pow(a, e / 2);
        return (b * b) % p;
    }
}

int Dot(int i, int j) {
    int k;
    int dot = 0;
    for (k = 0; k <= d - 1; k++)
        dot = (dot + Mat[i][k] * Mat[j][k]) % p;
    return dot;
}

void Add(int i, int j) {
    int k;
    for (k = 0; k <= d - 1; k++)
        Mat[j][k] = (Mat[j][k] + Mat[i][k]) % p;
}

void Swap(int i, int j) {
    int k, t;
    for (k = 0; k <= d - 1; k++) {
        t = Mat[i][k];
        Mat[i][k] = Mat[j][k];
        Mat[j][k] = t;
    }
}

void Orthogonalize(int i, int j) {
    int proj, k;
    proj = (Dot(i, j) * Pow(Dot(i, i), p - 2)) % p;
    for (k = 0; k <= d - 1; k++)
        Mat[j][k] = (Mat[j][k] - ((proj * Mat[i][k]) % p) + p) % p;
}

void ReadMatrix() {
    int k, i;
    for (k = 0; k <= n - 1; k++)
        for (i = 0; i <= d - 1; i++)
            scanf("%d", &Mat[k][i]);
}

void PrintMatrix() {
    int k, i;
    for (k = 0; k <= n - 1; k++) {
        for (i = 0; i <= d - 1; i++)
            printf("%d ", Mat[k][i]);
        printf("\n");
    }
}

void GramSchmidt() {
    int k, i, j;
    for (k = 0; k <= n - 1; k++) {
        i = k;
        while (i < n) {
            if (Dot(i, i) != 0)
                break;
            i = i + 1;
        }

        if (i < n)
            Swap(k, i);
        else {
            i = k;
            while (i < n) {
                j = i + 1;
                while (j < n) {
                    if (Dot(i, j) != 0)
                        break;
                    j = j + 1;
                }

                if (j < n)
                    break;
                i = i + 1;
            }

            if (i < n) {
                Add(j, i);
                Swap(k, i);
            }
        }

        if (Dot(k, k) != 0) {
            i = k + 1;
            while (i < n) {
                Orthogonalize(k, i);
                i = i + 1;
            }
        }
    }
}

int main() {
    int k, i, j;
    scanf("%d%d", &d, &n);
    ReadMatrix();
    GramSchmidt();
    PrintMatrix();
    return 0;
}
